rfcs/023-images-endpoint/README.md

```diff
@@ -0,0 +1,48 @@
+# RFC 023: Images endpoint
+
+**Status:** :construction: Draft :building_construction:
+
+**Last updated:** 27/04/2020
+
+## Context and requirements
+
+- Images are now entities in the catalogue API
+- We want to be able to search for images, and to fetch specified images
+- The images endpoint should behave similarly to the works endpoint
+
+## Proposal
+
+The initial images endpoint will be in **beta**, and therefore will be liable to have breaking changes made without a version bump.
+
+### Image display model
+
+```
+{
+  "id": String,
+  "location": DigitalLocation,
+  "parentWork": String
+}
+```
+
+This is intentionally minimal at this point as we decide what fields an image should have.
+
+### Single image
+
+**`GET`** `/catalogue/v2/images/:image_id`
+
+*Query params*
+
+- `include` - optional list of fields, as with works endpoint
+
+### Multiple images
+
+**`GET`** `/catalogue/v2/images`
+
+*Query params*
+
+- `query` - text search query
+- `similarTo` - an image ID; returned images are sorted by visual similarity to this.
```

Comment:
This shouldn't be a separate query param, it should be an include. For naming I'd vote for `visuallySimilar`.

Reply:
I am not sure what you mean here / if my guess is right I don't think it's technically feasible:

- do you mean that if you `include` `visuallySimilar` then you get a list of visually similar images alongside each returned image? if so, how many / how do you specify how many?

- @harrisonpim would it be possible to include such a list or would getting the list of visually similar images for each image require 1 query per result?

Reply:
Yes, that's what I meant. It might not be feasible or desirable.

Reply:
You're talking me round to it being a query parameter, I think I was wrong before.

Reply:
If it is an include (meaning when we ask for it, we get a list of visually similarly images on the given work) it might make sense only including it on the single work endpoint? If visually similar images are retrieved dynamically (I'm guessing via the feature vector stuff) intuitively it would be pretty slow for say a list of 10 works each getting 4 visually similar images.

Reply:
Btw I hadn't refreshed my browser before I wrote my comment so I missed everything after the initial comment

Reply:
> only including it on the single image endpoint

That's a good suggestion which I am kind of convinced by. The difference is between searching for images (sorted by similarity to a given one) and between asking for a specific image, as well as those most similar to it - and I'm not sure which of those I prefer. Technically there's basically no difference.

What do other people think?

Reply:
The suggestion I made of only having visually similar images on the single image endpoint would be a retrieval by ID and a single query. I don't reckon you could do the list of images and visually similar images too in a single query very easily.

Reply:
> The difference is between searching for images (sorted by similarity to a given one) and between asking for a specific image, as well as those most similar to it - and I'm not sure which of those I prefer. Technically there's basically no difference.

If you were always going to display a number of visually similar images on the single image endpoint the include makes sense, if it is something you might want todo async in response to some user input (after having received the image itself) then the original query parameter suggestion probably makes more sense

Reply:
> I don't reckon you could do the list of images and visually similar images too in a single query very easily.

Yep agreed. The difference I'm talking about is only in terms of how you make the single-id search, either:

- `/images/[id]?includes=visuallySimilar` or
- `/images/?visuallySimilarTo=[id]`

(obviously they'd have different return types too)

Reply:
I re-read @jtweed's comment below, have been convinced by the former of these options

Reply:
`/images/[id]?includes=visuallySimilar` feels more appropriate to me, and fits with @jtweed's "entity-centric" comment below

Reply:
Though I was actually becoming more convinced by the latter as it's more flexible and lets us decide how many to return. At the expense of requiring another API call to populate that component, but I think that's ok and arguably preferable as we only need visually similar images for those that are clicked on.